Refactor Twig environments and extensions

admdly opened the pull request:
Summary:

- Refactored Twig extensions into dedicated extension classes
- Added TwigFactory for centralised environment configuration
- Added markdown service and improved URL filter handling
- Added sandboxed Twig environment for email templates
- Added comprehensive tests for the sandbox

Migration guide and related docs will be updated prior to next release.

admdly committed:
Update Twig components and migrate to FOSSBilling namespace

admdly committed:
Update Twig components and migrate to FOSSBilling namespace

admdly committed:
Replace Custom Twig components

Replace custom Twig components with native Twig components, remove unused custom Twig components, and move functionality to macros where appropriate.

m

Update partial_extensions.html.twig

admdly committed:
Replace Custom Twig components

Replace custom Twig components with native Twig components, remove unused custom Twig components, and move functionality to macros where appropriate.

m

Update partial_extensions.html.twig

admdly committed:
Update Currency module...

admdly commented on src/di.php:
```diff
@@ -771,39 +798,4 @@
- *
- * @return string
- */
-$di['parse_markdown'] = $di->protect(function (?string $content, bool $addAttributes = true) use ($di) {
-    $content ??= '';
-    $defaultAttributes = [];
-
-    // If we are defining the default attributes, build the list and add them to the config
-    if ($addAttributes) {
-        $attributes = $di['mod_service']('theme')->getDefaultMarkdownAttributes();
-        foreach ($attributes as $class => $classAttributes) {
-            $reflectionClass = new ReflectionClass($class);
-            $fqcn = $reflectionClass->getName();
-            $defaultAttributes[$fqcn] = $classAttributes;
-        }
-    }
-
-    $parser = new League\CommonMark\GithubFlavoredMarkdownConverter([
-        'html_input' => 'escape',
-        'allow_unsafe_links' => false,
-        'max_nesting_level' => 50,
-        'default_attributes' => $defaultAttributes,
-    ]);
-
-    if ($addAttributes) {
-        $parser->getEnvironment()->addExtension(new DefaultAttributesExtension());
-    }
-
-    return $parser->convert($content);
-});
```
It's draft for a reason 🥲 - but noted.

admdly committed:
Refactor Twig environment config and loading

admdly committed:
Refactor Twig environment config and loading

admdly committed:
Switch CSRF to cookie-based double-submit

admdly committed:
Add Api Twig extension and restore widget/svg helpers

admdly committed:
Add Api Twig extension and restore widget/svg helpers

admdly committed:
Cast language flags; improve JSON parsing

admdly committed:
Cast language flags; improve JSON parsing

admdly committed:
Add markdown DI service and use in Admin API

admdly committed:
Add markdown DI service and use in Admin API

admdly committed:
Refactor Twig extensions and markdown, simplify APIs

admdly committed:
Refactor Twig extensions and markdown, simplify APIs

admdly committed:
Add sandboxed email Twig env, update render APIs

admdly committed:
Fix pagination URL usage and menu class

admdly committed:
Fix pagination URL usage and menu class

admdly committed:
Fix HTML markup in email templates

admdly committed:
Use macro for avatars and status badges

admdly committed:
Fix Twig spacing and pagination URL filter order

admdly committed:
Fix Twig spacing and pagination URL filter order

admdly committed:
Use `current_theme` global and refactor `period_title`

admdly committed:
Allow markdown and date in email Twig sandbox

admdly committed:
Update email template examples

admdly committed:
Fix Twig URLs, quoting, and logo alt text

admdly committed:
Fix Twig templates: URLs, i18n, search params

admdly committed:
Use `templates/*` directories for module templates

admdly committed:
Tighten Twig sandbox and add tests

admdly committed:
Merge branch 'main' into refactor/twig

admdly committed:
Add Twig environment and template docs to `AGENTS.md`

admdly committed:
Merge branch 'refactor/twig' of https://github.com/admdly/FOSSBilling into pr/2752

admdly committed:
Use Symfony Filesystem/Finder and assorted fixes

admdly committed:
Merge branch 'main' into refactor/twig

admdly committed:
Merge branch 'main' into refactor/twig

yagiz-dev commented:
How are we going to handle updates from a previous version? I couldn't see anything related in the UpdatePatcher.

I'm also not sure if we currently can run the patcher before having the templates renamed, it probably would throw an exception as templates are now expected to be in their new paths.

yagiz-dev commented on src/modules/Index/templates/client/mod_index_dashboard.html.twig:
```diff
@@ -209,26 +210,24 @@
-                            <div class="d-flex gap-2">
-                                <a class="btn btn-primary" href="{{ '/login'|link }}">{{ 'Login'|trans }}</a>
-                                <a class="btn btn-dark" href="{{ '/signup'|link }}">{{ 'Register'|trans }}</a>
-                            </div>
```
Accidentally removed </div> breaks homepage layout.

yagiz-dev reviewed (commented):
This also breaks existing email templates. We also should ideally automatically migrate these, at the very least the default ones.

<img width="1860" height="835" alt="Image" src="https://github.com/user-attachments/assets/212688e4-ae09-400c-8d6d-a4cae7ee6b68" />

yagiz-dev reviewed (commented):
Not sure if it's caused by this PR but the "Try your popup" button in Orderbutton is broken now.

<img width="1867" height="776" alt="Image" src="https://github.com/user-attachments/assets/f51fc88c-32a2-43a2-bbf3-a9602349f480" />

URL: `/admin/extension/settings/orderbutton`

yagiz-dev commented on src/modules/Product/templates/admin/mod_product_manage.html.twig:
```diff
@@ -303,19 +303,19 @@
 
             <div class="card-footer text-center">
-                <a href="{{ 'order'|link }}/{{ product.slug }}" class="btn btn-primary" target="_blank">
+                <a href='{{ "order/#{product.slug}"|url }}' class="btn btn-primary" target="_blank">
```
Should have been a client area URL.

yagiz-dev commented on src/modules/Embed/templates/admin/mod_embed_index.html.twig:
```diff
@@ -31,7 +31,7 @@
                 <div class="input-group input-group-flat">
                     <input id="domain-checker-form" type="text" class="form-control" autocomplete="off" readonly
-                           value='<iframe width="360" height="75" src="{{ 'embed/domainchecker'|link }}" frameborder="0"></iframe>'>
+                           value='<iframe width="360" height="75" src="{{ 'embed/domainchecker'|url }}" frameborder="0"></iframe>'>
```
Should have been a client area URL.

yagiz-dev commented on src/library/FOSSBilling/Twig/Extension/LegacyExtension.php:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    #[AsTwigFilter('ip_country_name')]
+    public function ipCountryName(string $ip): string
```
Are we certain that this isn't ever called with null? One example that came to my mind is the demo mode where we don't store IP addresses.

A null `$ip` would now throw a TypeError.

yagiz-dev commented on src/modules/Invoice/templates/admin/mod_invoice_invoice.html.twig:
```diff
@@ -115,28 +115,28 @@
 
             <div class="card-footer text-center">
-                <a href="{{ 'invoice'|link }}/{{ invoice.hash }}" class="btn btn-primary" target="_blank">
+                <a href='{{ "invoice/#{invoice.hash}"|url }}' class="btn btn-primary" target="_blank">
```
Should have been a client URL.

admdly committed:
Merge branch 'main' into pr/2752

admdly committed:
Merge branch 'refactor/twig' of https://github.com/admdly/FOSSBilling into pr/2752

admdly committed:
Fix ticket reply link fragment concatenation

admdly committed:
Close missing divs in dashboard template

admdly commented on src/modules/Index/templates/client/mod_index_dashboard.html.twig:
```diff
@@ -209,26 +210,24 @@
-                            <p>{{ 'Order new products and services'|trans }}</p>
-                            <div class="d-flex gap-2">
-                                <a class="btn btn-dark" href="{{ '/order'|link }}">{{ 'Order'|trans }}</a>
-                            </div>
```
Restored accidentally removed `</div>`.

admdly committed:
Use detailed `toApiArray` for client

admdly committed:
Fix module name resolution for email templates

admdly committed:
Regenerate `composer.lock`

admdly committed:
Use `markdown_to_html` filter in admin macro

admdly committed:
Handle null price/quantity and rename test method

admdly committed:
Use `Tools::isHTTPS()` for HTTPS checks

admdly committed:
Merge branch 'main' into pr/2752

admdly committed:
Merge branch 'main' into pr/2752

admdly committed:
Merge branch 'main' into pr/2752

admdly committed:
Merge branch 'main' into pr/2752

admdly committed:
Merge branch 'main' into pr/2752

admdly committed:
Merge branch 'main' into pr/2752

admdly committed:
Add 'area' param to `url` filter and update templates

admdly committed:
Merge branch 'main' into pr/2752

admdly committed:
Use client area URLs in admin templates

admdly committed:
Merge branch 'main' into pr/2752

admdly commented on src/modules/Support/templates/client/mod_support_contact_us_conversation.html.twig:
```diff
@@ -84,7 +84,7 @@
         <div class="card-footer" id="reply-to">
             {% if ticket.status != 'closed' %}
-                <form method="post" action="{{ 'api/guest/support/ticket_reply'|link }}" {{ fb_api_form({ reload: true }) }}>
+                <form method="post" action="{{ 'support/ticket_reply'|api_url }}" {{ fb_api_form({ reload: true }) }}>
```
Well noticed - updated to use guest API.

admdly committed:
Use client area and guest API URLs in templates

admdly committed:
Merge branch 'main' into pr/2752

admdly committed:
Merge branch 'main' into refactor/twig

admdly committed:
Remove client-side invoice delete feature

admdly committed:
Merge branch 'refactor/twig' of https://github.com/admdly/FOSSBilling into pr/2752

admdly committed:
Merge branch 'main' into pr/2752

admdly committed:
Create Twig env via `TwigFactory` for PDF rendering

admdly committed:
Merge branch 'main' into pr/2752

admdly committed:
Merge branch 'main' into pr/2752

admdly committed:
Merge branch 'main' into refactor/twig

admdly committed:
Merge branch 'main' into refactor/twig

admdly committed:
Merge branch 'main' into pr/2752

admdly committed:
Merge branch 'refactor/twig' of https://github.com/admdly/FOSSBilling into pr/2752

admdly committed:
Add embedded orderbutton JS and update render

admdly committed:
Add 'srclink' to cspell allowlist

admdly committed:
Merge branch 'main' into pr/2752

admdly committed:
Merge branch 'main' into pr/2752

admdly committed:
Use `renderEmailTplString` for email rendering

admdly committed:
Bump twig/markdown-extra to v3.24

admdly committed:
Use `markdown_to_html` for privacy policy

admdly committed:
Merge branch 'main' into pr/2752

admdly commented:
> Not sure if it's caused by this PR but the "Try your popup" button in Orderbutton is broken now.
> 
> <img alt="Image" width="1867" height="776" src="https://private-user-images.githubusercontent.com/35808275/563715985-f51fc88c-32a2-43a2-bbf3-a9602349f480.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzU0OTM2ODIsIm5iZiI6MTc3NTQ5MzM4MiwicGF0aCI6Ii8zNTgwODI3NS81NjM3MTU5ODUtZjUxZmM4OGMtMzJhMi00M2EyLWJiZjMtYTk2MDIzNDlmNDgwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA0MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNDA2VDE2MzYyMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNjYWZiYmQ2MzhmM2VhOTVkNTFkY2U1ZDdmODZiYTdjZWVkYjE1NzI4NTVhYzM2NDAyM2RiZjgwOTliYTA3MDQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.CmyerJiHh8HrYy_A9ktYTPngh6RdAUamXOttpdY979E">
> URL: `/admin/extension/settings/orderbutton`

As of testing this branch today, this seems to be working fine - perhaps a transient/fixed issue?

admdly commented on src/library/FOSSBilling/Twig/Extension/ApiExtension.php:
```diff
@@ -0,0 +1,90 @@
+    {
+        $json = json_encode($config, JSON_THROW_ON_ERROR | JSON_UNESCAPED_SLASHES);
+
+        return 'data-fb-api=\'' . $json . '\'';
```
Agreed - single quotes escaped.

admdly commented on src/modules/Staff/templates/email/mod_staff_client_signup.html.twig:
```diff
@@ -39,7 +39,7 @@
 	<p><strong>{{ c.first_name }} {{ c.last_name }}</strong> has just signed up with {{ guest.system_company.name }}</p>
 
-    <p><a href="{{'client/manage'|alink}}/{{c.id}}" target="_blank">Manage client.</a>
+    <p><a href='{{ "client/manage/#{c.id}"|url(detect_app_area: false) }}' target="_blank">Manage client.</a>
```
Updated elsewhere - should resolve comment.

admdly committed:
Escape API JSON, handle null IP, client URLs

admdly committed:
Merge branch 'main' into refactor/twig

admdly committed:
Merge branch 'main' into pr/2752

admdly committed:
Remove initial admin creation flow

admdly committed:
Handle HTTP failures and make tests robust

admdly commented:
> How are we going to handle updates from a previous version? I couldn't see anything related in the UpdatePatcher.
> 
> I'm also not sure if we currently can run the patcher before having the templates renamed, it probably would throw an exception as templates are now expected to be in their new paths.

Working on this now - hopefully we can handle most cases without manual intervention.

admdly committed:
Merge branch 'main' into pr/2752

admdly committed:
Use `api_url` helper in admin templates

admdly committed:
Use `url` filter for footer link 5

admdly commented:
> How are we going to handle updates from a previous version? I couldn't see anything related in the UpdatePatcher.
> 
> I'm also not sure if we currently can run the patcher before having the templates renamed, it probably would throw an exception as templates are now expected to be in their new paths.

Looking at this, the new structure is extracted before the patcher runs, so everything should be in place to avoid an exception. However, some handling is needed to handle/remove legacy files, so I've updated the patcher. 

For other, non-supported modules, this is a breaking release and we're pre-v1; I'm not sure how much we can try to handle these cases without becoming overwhelmed.

admdly committed:
Add patch53 to remove legacy html templates

admdly committed:
Bump setting last_patch to 53

admdly commented:
> This also breaks existing email templates. We also should ideally automatically migrate these, at the very least the default ones.
> 
> <img alt="Image" width="1860" height="835" src="https://private-user-images.githubusercontent.com/35808275/563715620-212688e4-ae09-400c-8d6d-a4cae7ee6b68.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzU4NjI3NjcsIm5iZiI6MTc3NTg2MjQ2NywicGF0aCI6Ii8zNTgwODI3NS81NjM3MTU2MjAtMjEyNjg4ZTQtYWUwOS00MDBjLThkNmQtYTRjYWU3ZWU2YjY4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA0MTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNDEwVDIzMDc0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIxMDA0NDg2NGEwYzBjOTRjNzNmNzhhZDI1NTUwN2Y0NTY1ZTExZmRmOGE0MTRmZWU5YzliMTMwYTE1ZTQ4MjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.aktJPPlqwrELBDNEcbJq25OD-101Yh5qSLZWaA6pQ5U">

I've added a patch to update these where they haven't changed from default, but migrating modified templates would be much more complex. This is something we should likely highlight in the release notes, as a breaking release, and develop some migration docs around.

admdly committed:
Add patch54 to restore legacy email templates

admdly committed:
Use renderEmailTplString for email templating

admdly committed:
Remove redundant else/endif in order template

admdly committed:
Merge branch 'main' into pr/2752

admdly committed:
Merge branch 'main' into refactor/twig

admdly commented:
@copilot apply changes based on the comments in [this thread](https://github.com/FOSSBilling/FOSSBilling/pull/2752#pullrequestreview-4093025602)

admdly committed:
Add file-hash cache buster for asset URLs

admdly committed:
Merge branch 'main' into pr/2752

admdly committed:
Remove raw Twig filter and handle missing modules dir

admdly committed:
Return '-' for future timestamps

admdly committed:
Register and reuse TwigFactory and add CSRF cookie

admdly committed:
Adjust API URL helper and add Twig error handling

admdly commented on src/modules/System/Service.php:
```diff
@@ -479,6 +481,37 @@ public function createTemplateFromString($tpl, $try_render, $vars)
+    public function renderEmailTplString(string $tpl, array $vars): string
+    {
+        $twigFactory = new \FOSSBilling\Twig\TwigFactory($this->di);
+        $twig = $twigFactory->createEmailEnvironment();
+
+        try {
+            $template = $twig->createTemplate($tpl);
+
+            return $template->render($vars);
+        } catch (\Twig\Sandbox\SecurityError $e) {
+            $this->di['logger']->setChannel('security')->warning('Email template sandbox violation', [
+                'error' => $e->getMessage(),
+            ]);
+
+            throw new \FOSSBilling\InformationException('Email template contains disallowed Twig syntax: ' . $e->getMessage());
+        } catch (\Twig\Error\SyntaxError $e) {
+            throw new \FOSSBilling\InformationException('Email template syntax error: ' . $e->getMessage());
+        }
```
Updated to add `try...catch`.

admdly commented on src/modules/System/templates/admin/mod_system_update.html.twig:
```diff
@@ -23,22 +23,20 @@
         <div class="card-footer">
             {% if admin.system_update_available or update_info['branch'] == 'preview' %}
-                <a class="btn btn-primary" href="{{ 'api/admin/system/update_core'|link }}"
+                <a class="btn btn-primary" href="{{ 'system/update_core'|api_url }}"
```
Updated to remove unnecessary third parameter.

admdly committed:
Merge branch 'main' into pr/2752

admdly committed:
Update composer.lock and package-lock.json

admdly committed:
Merge branch 'main' into pr/2752

admdly committed:
Add twig_factory to ServiceTest DI

admdly committed:
Use url filter for link_page in template

admdly commented on src/modules/Spamchecker/Service.php:
```diff
@@ -191,14 +191,19 @@ public function isTemp(\Box_Event $event): void
+            $file_contents = $response->getContent(false);
+        } catch (\Exception) {
+            return false;
+        }
```
Not applicable for PR.

admdly committed:
Merge branch 'main' into pr/2752

admdly committed:
Composer - update deps and add markdown-extra

admdly committed:
Merge branch 'main' into pr/2752

admdly committed:
Remove docblocks and inline comments in `TwigFactory`

yagiz-dev commented on src/modules/Orderbutton/templates/client/mod_orderbutton_choose_product.html.twig:
```diff
@@ -10,10 +10,10 @@
             {% for i, category in guest.product_category_get_list.list %}
                 <h4 {% if i > 0 %} class="mt-3" {% endif %}>{{ category.title }}</h4>
-                <span>{{ category.description|markdown }}</span>
+                <span>{{ category.description|markdown_to_html }}</span>
```
@admdly I can still replicate it. Passing null to markdown_to_html isn't allowed and category descriptions are nullable.

We could wrap this with a check like this or perhaps tweak markdown_to_html to accept nulls?

```twig
{% if category.description is not empty %}
    <span>{{ category.description|markdown_to_html }}</span>
{% endif %}
```

admdly replied:
I've updated this to `<span>{{ (category.description ?? '')|markdown_to_html }}</span>`, which should resolve the problem.

yagiz-dev commented on src/library/FOSSBilling/Twig/Extension/FOSSBillingExtension.php:
```diff
@@ -0,0 +1,268 @@
+<?php
```
The new `has_permission` function from #3396 seems yet to be ported here.

admdly replied:
This seems to have been lost when `TwigExtensions.php` was dropped merging `main` - I've ported it over.

yagiz-dev commented on src/modules/System/templates/admin/mod_system_settings.html.twig:
Removed `link` function still in use here in line 660.

yagiz-dev replied:
Also broken duplicate form from line 660 onwards. Likely happened when merging main.

admdly replied:
Looks like both were from a broken merge in of `main` - thanks for pointing these out - fixed. 🙂

admdly committed:
Merge branch 'main' into pr/2752

admdly committed:
Add patches 53-55 and update `last_patch`

admdly committed:
Add `has_permission` Twig function; exclude `href`

admdly committed:
Remove duplicate network interface form

admdly committed:
Guard removal of legacy dirs; add network input

admdly committed:
Merge branch 'main' into pr/2752

admdly committed:
Handle null category descriptions in template

yagiz-dev commented:
I think we should find a better solution to the nullable markdown fields. The same thing happens for addon product descriptions and probably other places.

Can we make markdown_to_html return an empty string when it's passed null? Maybe we can override it with a custom runtime like this:

```php

<?php

declare(strict_types=1);
/**
 * Copyright 2022-2025 FOSSBilling.
 * SPDX-License-Identifier: Apache-2.0.
 *
 * @copyright FOSSBilling (https://www.fossbilling.org)
 * @license http://www.apache.org/licenses/LICENSE-2.0 Apache-2.0
 */

namespace FOSSBilling\Twig\Markdown;

use Twig\Extra\Markdown\MarkdownRuntime;

class FOSSBillingMarkdownRuntime extends MarkdownRuntime
{
    public function convert($body): string
    {
        if ($body === null || $body === '') {
            return '';
        }

        return parent::convert((string) $body);
    }
}
```

I think this would also make it consistent with our existing markdown engine.

admdly committed:
Merge branch 'main' into pr/2752

admdly committed:
Add Twig adapter env and sandbox policy

admdly committed:
Add 'striptags' to cspell word list

admdly committed:
Update composer.lock (deps updated/removed)

admdly committed:
Merge branch 'main' into pr/2752

admdly committed:
Sanitize custom payment adapter templates

admdly committed:
Remove redundant legacy API tests

admdly committed:
Bump `last_patch` setting to 56

admdly committed:
Coalesce null before `markdown_to_html`

admdly commented:
> I think we should find a better solution to the nullable markdown fields. The same thing happens for addon product descriptions and probably other places.
> 
> Can we make markdown_to_html return an empty string when it's passed null? Maybe we can override it with a custom runtime like this:
> 
> ```
> <?php
> 
> declare(strict_types=1);
> /**
>  * Copyright 2022-2025 FOSSBilling.
>  * SPDX-License-Identifier: Apache-2.0.
>  *
>  * @copyright FOSSBilling (https://www.fossbilling.org)
>  * @license http://www.apache.org/licenses/LICENSE-2.0 Apache-2.0
>  */
> 
> namespace FOSSBilling\Twig\Markdown;
> 
> use Twig\Extra\Markdown\MarkdownRuntime;
> 
> class FOSSBillingMarkdownRuntime extends MarkdownRuntime
> {
>     public function convert($body): string
>     {
>         if ($body === null || $body === '') {
>             return '';
>         }
> 
>         return parent::convert((string) $body);
>     }
> }
> ```
> 
> I think this would also make it consistent with our existing markdown engine.

I'd rather avoid overriding default functions if we can as it just creates more work in the longer term as we have to maintain it. I've been through and added null checks before any calls to `markdown_to_html`, and we should probably avoid returning `null` for such values going forward as things are updated.

yagiz-dev commented:
> We should probably avoid returning null for such values going forward as things are updated.

This makes the most sense. For now, we should be fine with adding null checks.

I'll tackle this with another PR in the future.

admdly committed:
Merge branch 'main' into refactor/twig

admdly committed:
Merge branch 'main' into pr/2752

admdly committed:
Merge branch 'refactor/twig' of https://github.com/admdly/FOSSBilling into pr/2752

admdly committed:
Move update patcher to system CLI command

admdly committed:
Merge branch 'main' into refactor/twig

admdly committed:
Merge branch 'main' into pr/2752

admdly committed:
Merge branch 'refactor/twig' of https://github.com/admdly/FOSSBilling into pr/2752

admdly committed:
Merge branch 'main' into pr/2752

Resolve merge conflicts between the Twig refactoring branch and main:
- Template conflicts: combine new Twig syntax with main's Antispam features
  and permission guards
- UpdatePatcher: renumber branch patches to 56-58 to avoid collision with
  main's patches 54-55
- cspell.json: include words from both sides
- Email API: keep main's permission check with branch's DI style
- AdminTest: use main's retry loop for reliability

admdly committed:
Remove stray `endif` from recaptcha macro

admdly committed:
Merge branch 'main' into pr/2752

# Conflicts:
#	composer.lock
#	src/install/sql/content.sql
#	src/library/Box/TwigExtensions.php
#	src/library/FOSSBilling/UpdatePatcher.php
#	src/modules/News/templates/client/mod_news_post.html.twig
#	src/modules/Staff/templates/admin/mod_staff_password_reset.html.twig
#	src/modules/Support/templates/client/mod_support_ticket.html.twig
#	src/modules/System/templates/admin/mod_system_settings.html.twig
#	src/modules/Theme/templates/admin/mod_theme_settings.html.twig
#	src/themes/admin_default/html/macro_functions.html.twig
#	tests-legacy/modules/Api/ServiceTest.php
#	tests-legacy/modules/Cron/Api/GuestTest.php
#	tests-legacy/modules/Staff/Api/GuestTest.php
#	tests-legacy/modules/System/Api/GuestTest.php

admdly committed:
Update composer.lock

admdly committed:
Fix extra closing brace in Api ServiceTest

admdly committed:
Fix test errors from main merge

- Remove testGetRequestCount tests (method moved to RateLimiter)
- Remove testGetDi from Staff and System GuestTest (protected method)
- Fix IP error message in testLoginCheckIpException
- Add rate_limiter mock to Staff GuestTest tests that need it

admdly committed:
Fix remaining Staff GuestTest issues

- Add extension service mock for captcha check
- Fix error message from 'User not found' to 'Admin not found'

admdly committed:
Fix Staff GuestTest: update query and add logger mock

admdly committed:
Fix Staff GuestTest: set admin status to active for password reset flow

admdly committed:
Fix stale API calls and inverted required condition

- Notification: replace removed system_string_render with direct output
- Antispam settings: replace |alink/|link with |url/|api_url filters
- build_form macro: fix inverted required condition for non-textarea inputs

admdly committed:
Sanitize JS/data URIs and disable buttons by permission

admdly committed:
Use client login/profile links in support emails

admdly committed:
Encode fb-api JSON and tighten URL sanitization

admdly committed:
Prefer session auth for API key methods

admdly committed:
Merge branch 'main' into pr/2752

admdly committed:
Remove stray endif and fix template formatting

admdly committed:
Add CSRF token to CSV exports; fix template

admdly committed:
Remove stray endif in currency template

admdly committed:
Make invoice PDF generation return Response

admdly committed:
Merge branch 'main' into pr/2752

# Conflicts:
#	src/modules/Invoice/Service.php
#	tests-legacy/modules/Invoice/ServiceTest.php

admdly committed:
Fix Twig filters, templates, and form handling

admdly committed:
Remove readonly from anonymous class in test

admdly committed:
Inline DebugBar CSS/JS in renderHead

admdly committed:
Merge branch 'main' into refactor/twig

admdly committed:
Add Bootstrap Tab to global bootstrap

admdly committed:
Capitalize template UI strings

admdly committed:
Use CKEditor API for reply quotes

admdly committed:
Remove WYSIWYG module and move CKEditor to themes

admdly committed:
Bump `last_patch` in install SQL to 59

admdly committed:
Remove Spamchecker module

admdly committed:
Run domain management script only for active orders

admdly committed:
Standardize UI text capitalization & minor fixes

admdly committed:
Remove Spamchecker legacy tests; fix period test

admdly committed:
Rename Spamchecker tests to Antispam

admdly committed:
Restore antispam config in tests; remove fixture

admdly committed:
Refactor Antispam Guest tests for config checks

admdly committed:
Update theme build docs; remove Wysiwyg module

admdly committed:
Remove spamchecker from cspell and tweak comment

admdly committed:
Add spamchecker to cspell word list

admdly committed:
Capitalize 'Update Graphs' button label
